compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/cyclefind/GlobalVariableRefAnalyzer.java

```diff
@@ -0,0 +1,356 @@
+
+        // Taking a copy as we need to modify the original list.
+        List<BSymbol> providersProviders = new ArrayList<>(globalNodeDependsOn.get(provider));
+        for (BSymbol prov : providersProviders) {
+            pruneFunctions(provider, prov, globalNodeDependsOn, visited);
+        }
+    }
+
+    private void addDependenciesDependencies(LinkedList<BSymbol> dependencies, Set<BSymbol> sorted) {
+        // For each dependency if they satisfy their dependencies in sorted list, then add them to sorted list.
+        ArrayList<BSymbol> depCopy = new ArrayList<>(dependencies);
+        for (BSymbol dep : depCopy) {
+            Set<BSymbol> depsDependencies = globalNodeDependsOn.getOrDefault(dep, new LinkedHashSet<>());
+            if (!depsDependencies.isEmpty() && sorted.containsAll(depsDependencies)) {
+                moveAndAppendToSortedList(dep, dependencies, sorted);
+            }
+        }
+    }
+
+    private void projectSortToTopLevelNodesList() {
+        // Swap global variable nodes in 'topLevelNodes' list to reflect sorted global variables.
+        List<Integer> topLevelPositions = new ArrayList<>();
+        for (BLangSimpleVariable globalVar : pkgNode.globalVars) {
+            topLevelPositions.add(pkgNode.topLevelNodes.indexOf(globalVar));
+        }
+        topLevelPositions.sort(Comparator.comparingInt(i -> i));
+        for (int i = 0; i < topLevelPositions.size(); i++) {
+            Integer targetIndex = topLevelPositions.get(i);
+            pkgNode.topLevelNodes.set(targetIndex, pkgNode.globalVars.get(i));
+        }
+    }
+
+    private void projectSortToGlobalVarsList(Set<BSymbol> sorted) {
+        Map<BSymbol, BLangSimpleVariable> varMap = this.pkgNode.globalVars.stream()
+                .collect(Collectors.toMap(k -> k.symbol, k -> k));
+
+        List<BLangSimpleVariable> sortedGlobalVars = sorted.stream()
+                .filter(varMap::containsKey)
+                .map(varMap::get)
+                .collect(Collectors.toList());
+
+        if (sortedGlobalVars.size() != this.pkgNode.globalVars.size()) {
+            List<BLangSimpleVariable> symbolLessGlobalVars = this.pkgNode.globalVars.stream()
+                    .filter(g -> g.symbol == null)
+                    .collect(Collectors.toList());
+            sortedGlobalVars.addAll(symbolLessGlobalVars);
+        }
+        this.pkgNode.globalVars.clear();
+        this.pkgNode.globalVars.addAll(sortedGlobalVars);
+    }
+
+    private List<BSymbol> getGlobalVariablesAndDependentFunctions() {
+        List<BSymbol> globalVars = this.pkgNode.globalVars.stream()
+                .filter(v -> v.symbol != null)
+                .map(v -> v.symbol)
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        List<BSymbol> funcs = this.globalNodeDependsOn.keySet().stream()
+                .filter(symbol -> !globalVars.contains(symbol))
+                .collect(Collectors.toList());
+
+        funcs.addAll(globalVars);
```

Comment:
Without going for a full blown for loop I don't see any perf gain with pre-allocated list.

Reply:
Collectors.toList() creates intermediate lists which we are not using anyway. And addAll() will internally traverse through the list once more.

Both of these can be replaced by a simple list.add()

Reply:
What Collectors.toList() create is a array list and then we are addAll to that list, because it's a array list addAll just resize the internal array and perform a arrayCopy.

We are talking about line 221 right?

Reply:
My point is,

```ballerina
List<BSymbol> funcs = this.globalNodeDependsOn.keySet().stream()
            .filter(symbol -> !globalVars.contains(symbol))
            .collect(Collectors.toList());

funcs.addAll(globalVars);
```
can be replaced by
```ballerina
globalNodeDependsOn.keySet().forEach(symbol -> globalVars.add(symbol));
```

Reply:
What I want is this:
```java
        List<BSymbol> dependents = new ArrayList<>();

        for (BSymbol s : globalNodeDependsOn.keySet()) {
            if (s.tag == SymTag.FUNCTION) {
                dependents.add(s);
            }
        }

        for (BLangSimpleVariable var : this.pkgNode.globalVars) {
            if (var.symbol != null) {
                dependents.add(var.symbol);
            }
        }

        return dependents;
```

Reply:
If `globalNodeDependsOn` contains only function-symbols and global var symbols, I think both your code and what I suggests gives the same output :)

Reply:
At the moment globalNodeDependsOn.keySet() contain functions and variables (global) and we will need to add constants in the future to support them.